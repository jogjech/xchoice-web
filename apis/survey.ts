import axios from "axios";
import {
  Survey,
  SurveyResponse,
  SurveyMetadata,
  SurveyReport,
} from "../models/survey";

// For get requests, we can use SWR https://swr.vercel.app/getting-started

interface PostSurveyResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyId?: string;
}

interface PostSurveyResponseResult {
  error?: {
    message: string;
  };
  isError: boolean;
  slug?: string; // slug is the short string which can be used to view / edit this surey respnose
}

interface GetSurveyResult {
  error?: {
    message: string;
  };
  isError: boolean;
  survey?: Survey;
}

interface GetSurveyReportResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyReport?: SurveyReport;
}

interface GetSurveyResponseResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyResponse?: SurveyResponse;
}

interface FindSurveysResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyMetadataList?: SurveyMetadata[];
}

const postSurvey = async (survey: Object): Promise<PostSurveyResult> => {
  console.log("Posting survey", survey);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  return {
    isError: false,
    surveyId: "1",
  };
  // return {
  //   isError: true,
  //   error: {
  //     message: "Unknown",
  //   },
  //   surveyId: "1",
  // };
};

const postSurveyResponse = async (
  surveyResponse: SurveyResponse
): Promise<PostSurveyResponseResult> => {
  console.log("Posting survey response", surveyResponse);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  if (surveyResponse.surveyId === "1") {
    return {
      isError: false,
      slug: "fasd07fhfak",
    };
  } else {
    return {
      isError: true,
      error: {
        message: "Unknown",
      },
    };
  }
};

const getSurveyResponse = async (
  slug: string
): Promise<GetSurveyResponseResult> => {
  console.log("Getting survey response", slug);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  if (slug === "fasd07fhfak") {
    return {
      isError: false,
      surveyResponse: {
        surveyId: "1",
        selections: [1, 3],
      },
    };
  } else {
    return {
      isError: true,
      error: {
        message: "Unknown",
      },
    };
  }
};

const getSurvey = async (surveyId: string): Promise<GetSurveyResult> => {
  console.log("Getting survey", surveyId);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  if (surveyId === "1") {
    return {
      isError: false,
      survey: {
        published: true,
        surveyTitle: "Personal survey",
        questions: [
          {
            questionTitle: "Your age",
            choices: [
              {
                text: "<18",
              },
              {
                text: "18-30",
              },
              {
                text: "31-60",
              },
              {
                text: ">60",
              },
            ],
          },
          {
            questionTitle: "Your job family",
            choices: [
              {
                text: "Engineering",
              },
              {
                text: "Finance",
              },
              {
                text: "Sales",
              },
              {
                text: "Others",
              },
            ],
          },
        ],
      },
    };
  } else {
    return {
      isError: true,
      error: {
        message: "Cannot find survey",
      },
    };
  }
};

const getSurveyReport = async (
  surveyId: string
): Promise<GetSurveyReportResult> => {
  console.log(`Getting survey report for survey ${surveyId}`);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  if (surveyId === "1") {
    return {
      isError: false,
      surveyReport: {
        published: true,
        surveyId: surveyId,
        responses: 11,
        surveyTitle: "Personal survey",
        questions: [
          {
            questionTitle: "Your age",
            choices: [
              {
                text: "<18",
                selections: 3,
              },
              {
                text: "18-30",
                selections: 3,
              },
              {
                text: "31-60",
                selections: 4,
              },
              {
                text: ">60",
                selections: 1,
              },
            ],
          },
          {
            questionTitle: "Your job family",
            choices: [
              {
                text: "Engineering",
                selections: 5,
              },
              {
                text: "Finance",
                selections: 4,
              },
              {
                text: "Sales",
                selections: 1,
              },
              {
                text: "Others",
                selections: 1,
              },
            ],
          },
        ],
      },
    };
  } else {
    return {
      isError: true,
      error: {
        message: "Cannot load report for this survey",
      },
    };
  }
};

const deleteSurvey = async (surveyId: string) => {};

const findSurveys = async (userId: string): Promise<FindSurveysResult> => {
  console.log(`Finding surveys for user ${userId}`);
  await new Promise((resolve) => setTimeout(resolve, 1000));

  if (userId === "1") {
    return {
      isError: false,
      surveyMetadataList: [
        {
          surveyId: "1",
          responses: 0,
          published: false,
          userId: userId,
          title: "Personal survey",
        },
        {
          surveyId: "14",
          responses: 15052,
          published: true,
          userId: userId,
          title: "Work experience survey",
        },
        {
          surveyId: "12",
          responses: 0,
          published: false,
          userId: userId,
          title: "Personal survey",
        },
        {
          surveyId: "15",
          responses: 15052,
          published: true,
          userId: userId,
          title: "Work experience survey",
        },
        {
          surveyId: "17",
          responses: 0,
          published: false,
          userId: userId,
          title: "Personal survey",
        },
      ],
    };
  } else {
    return {
      isError: true,
      error: {
        message: "Cannot find surveys for this user",
      },
    };
  }
};

export {
  postSurvey,
  getSurvey,
  getSurveyReport,
  deleteSurvey,
  findSurveys,
  postSurveyResponse,
  getSurveyResponse,
};
