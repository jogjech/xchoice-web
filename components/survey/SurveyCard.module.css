.surveyCard {
  width: 20rem;
  height: 14rem;
  font-size: 1.5rem;
  margin-bottom: 2rem;
}
