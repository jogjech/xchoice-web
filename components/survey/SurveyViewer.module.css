.spinContainer {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.surveyTitle {
  font-size: 2rem;
  margin-bottom: 1rem;
}

.submitButton {
  margin: 1rem 0;
}
