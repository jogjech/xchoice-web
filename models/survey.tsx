export interface Choice {
  text: string;
}

export interface Question {
  questionTitle: string;
  choices: Choice[];
}

export interface Survey {
  surveyTitle: string;
  questions: Question[];
}

export interface SurveyResponse {
  surveyId: string;
  selections: number[];
}

export interface SurveyMetadata {
  surveyId: string;
  responses: number;
  published: boolean;
  userId: string;
  title: string;
}
