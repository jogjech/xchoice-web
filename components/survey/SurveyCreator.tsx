import React, { FunctionComponent, useState } from "react";
import { Form, Input, Button, Result, Alert } from "antd";
import QuestionCreator from "./question/QuestionCreator";
import { postSurvey } from "../../apis/survey";
import { useAuth0 } from "@auth0/auth0-react";
import { useRouter } from "next/router";

interface Props {}

const SurveyCreator: FunctionComponent<Props> = () => {
  const { getAccessTokenSilently } = useAuth0();

  const [form] = Form.useForm();
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const [surveyId, setSurveyId] = useState("");
  const [redirecting, setRedirecting] = useState(false);
  const router = useRouter();

  const handleFormFinish = async (formData: {
    surveyTitle: string;
    questions: [
      {
        questionTitle: string;
        choices: [{ text: string }];
      }
    ];
  }) => {
    console.log(formData);
    setPosting(true);
    setPostError(undefined);

    const accessToken = await getAccessTokenSilently();
    const result = await postSurvey(formData, accessToken);

    setPostError(result.error);
    setPosted(!result.isError);
    setSurveyId(result.surveyId);
    setPosting(false);
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title="Your survey is posted!"
          subTitle={`Survey ID: ${surveyId}.`}
          extra={[
            <Button
              type="primary"
              key="console"
              loading={redirecting}
              onClick={() => {
                setRedirecting(true);
                router.push(`/survey/${surveyId}`);
              }}
            >
              View my survey
            </Button>,
          ]}
        />
      ) : (
        <>
          <div>
            <p className="title">Create your survey</p>
          </div>
          <Form
            layout="vertical"
            form={form}
            initialValues={{
              questions: [
                {
                  choices: [
                    {
                      text: "",
                    },
                  ],
                },
              ],
            }}
            onFinish={handleFormFinish}
            autoComplete="off"
          >
            <p className="sub-title">Basic Info</p>
            <Form.Item
              label="Survey Title"
              name="surveyTitle"
              rules={[{ required: true, message: "Missing survey title" }]}
            >
              <Input placeholder="Survey Title"></Input>
            </Form.Item>
            <QuestionCreator></QuestionCreator>
            <Form.Item>
              <Button type="primary" htmlType="submit" loading={posting}>
                Submit
              </Button>
            </Form.Item>
          </Form>
          {postError && (
            <Alert
              message="Failed to create survey. Please try again."
              type="error"
            ></Alert>
          )}
        </>
      )}
    </>
  );
};

export default SurveyCreator;
