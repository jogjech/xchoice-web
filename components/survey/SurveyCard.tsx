import React, { FunctionComponent, useState } from "react";
import { Card, Statistic, Tag } from "antd";
import { useRouter } from "next/router";
import { SurveyMetadata } from "../../models/survey";
import { FileTextOutlined } from "@ant-design/icons";
import styles from "./SurveyCard.module.css";

interface Props {
  surveyMetadata: SurveyMetadata;
}

const SurveyCard: FunctionComponent<Props> = ({ surveyMetadata }) => {
  const router = useRouter();

  return (
    <>
      {!!surveyMetadata ? (
        <div>
          <Card
            title={surveyMetadata.title}
            hoverable
            bordered
            className={styles.surveyCard}
            headStyle={{
              fontSize: "1.4rem",
              backgroundColor: "rgba(26, 119, 213, 0.6)",
            }}
            onClick={() => {
              router.push(`/survey/${surveyMetadata.surveyId}`);
            }}
          >
            <div>
              <Statistic
                title="Responses"
                value={surveyMetadata.responses}
                prefix={<FileTextOutlined />}
              />
              {surveyMetadata.published ? (
                <Tag color="green">published</Tag>
              ) : (
                <Tag color="orange">unpublished</Tag>
              )}
            </div>
          </Card>
        </div>
      ) : (
        <div>
          <Card loading bordered className={styles.surveyCard}>
            <div></div>
          </Card>
        </div>
      )}
    </>
  );
};

export default SurveyCard;
