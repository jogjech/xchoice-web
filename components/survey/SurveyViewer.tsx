import React, { FunctionComponent, useState, useEffect } from "react";
import { Spin, Button, Result, Alert } from "antd";
import {
  getSurvey,
  postSurveyResponse,
  getSurveyResponse,
} from "../../apis/survey";
import { Survey } from "../../models/survey";
import QuestionViewer from "./question/QuestionViewer";
import { useRouter } from "next/router";
import styles from "./SurveyViewer.module.css";

interface Props {
  surveyId?: string;
  slug?: string;
}

const SurveyViewer: FunctionComponent<Props> = ({
  surveyId: inputSurveyId,
  slug: inputSlug,
}) => {
  // TODO: missing the handle of getSurvey and getSurveyResponse errors
  const [surveyId, setSurveyId] = useState<string>(undefined);
  const [survey, setSurvey] = useState<Survey>(undefined);
  const [loading, setLoading] = useState(true);
  const [validated, setValidated] = useState(false);
  const [selections, setSelections] = useState<number[]>([]);
  const [responseSlug, setResponseSlug] = useState<string>(undefined);
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const router = useRouter();

  useEffect(() => {
    const fetchSurvey = async () => {
      let tempSurveyId: string, tempSelections: number[];

      if (!!inputSurveyId) {
        tempSurveyId = inputSurveyId;
      } else {
        const result = await getSurveyResponse(inputSlug);
        tempSurveyId = result.surveyResponse.surveyId;
        tempSelections = result.surveyResponse.selections;
      }

      setSurveyId(tempSurveyId);

      const result = await getSurvey(tempSurveyId as string);

      setSelections(
        !!tempSelections
          ? tempSelections
          : Array.apply(null, Array(result.survey.questions.length))
      );

      setSurvey(result.survey);
      setLoading(false);
    };
    fetchSurvey();
  }, []);

  const handleSelectionChange = (questionIndex: number) => (
    newValue: number
  ) => {
    console.log(`Question ${questionIndex} received new value ${newValue}`);
    setValidated(false);
    setSelections((prevSelections) => {
      const newSelections = [...prevSelections];
      newSelections[questionIndex] = newValue;
      return newSelections;
    });
  };

  const handleSubmit = async () => {
    console.log("Submitting", selections);
    setPosting(true);
    setPostError(undefined);

    const validationPassed = validateSelections();
    if (validationPassed) {
      console.log("Validation passed");
      const result = await postSurveyResponse({
        surveyId: surveyId,
        selections: selections,
      });
      if (result.isError) {
        setPostError(result.error);
      } else {
        setResponseSlug(result.slug);
      }
      setPosted(!result.isError);
    } else {
      console.log("Validation failed");
    }
    setPosting(false);
  };

  const validateSelections: () => boolean = () => {
    setValidated(true);
    return selections.every((i) => i != null);
  };

  const handleViewResponseButtonClick = () => {
    const toPath = `/survey/response/${responseSlug}`;
    console.log(router);
    if (toPath === router.asPath) {
      router.reload();
    } else {
      router.push(`/survey/response/${responseSlug}`);
    }
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title={`Your response is recorded! Response id: ${responseSlug}`}
          subTitle={`Please use the response id or click the button below to view / edit your response.`}
          extra={[
            <Button
              type="primary"
              key="console"
              onClick={handleViewResponseButtonClick}
            >
              View my response
            </Button>,
          ]}
        />
      ) : loading ? (
        <div className={styles.spinContainer}>
          <Spin tip="Loading survey..."></Spin>
        </div>
      ) : (
        <div>
          <p className={styles.surveyTitle}>{survey.surveyTitle}</p>
          {survey.questions.map((question, index) => (
            <QuestionViewer
              question={question}
              key={index}
              selection={selections[index]}
              handleSelectionChange={handleSelectionChange(index)}
              submitWithNoSelection={validated && selections[index] == null}
            ></QuestionViewer>
          ))}
          <Button
            className={styles.submitButton}
            type="primary"
            onClick={handleSubmit}
            loading={posting}
          >
            Submit
          </Button>
          {postError && (
            <Alert
              message="Failed to post response. Please try again."
              type="error"
            ></Alert>
          )}
        </div>
      )}
    </>
  );
};

export default SurveyViewer;
