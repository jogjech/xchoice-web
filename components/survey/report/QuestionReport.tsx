import React, { FunctionComponent } from "react";
import { Question } from "../../../models/survey";
import { Progress } from "antd";
import styles from "./QuestionReport.module.css";

interface Props {
  question: Question;
}

const QuestionReport: FunctionComponent<Props> = ({ question }) => {
  const totalCount = question.choices
    .map((choice) => choice.selections)
    .reduce((a, b) => a + b, 0);

  return (
    <>
      <div>
        <p className={styles.questionTitle}>{question.questionTitle}</p>
      </div>
      <div className={styles.choices}>
        {question.choices.map((choice, index) => {
          return (
            <div key={index}>
              <span>{choice.text}</span>
              <Progress
                percent={Math.round((choice.selections / totalCount) * 100)}
              />
            </div>
          );
        })}
      </div>
    </>
  );
};

export default QuestionReport;
