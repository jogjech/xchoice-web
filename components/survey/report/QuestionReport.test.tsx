import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { Question } from "../../../models/survey";
import QuestionReport from "./QuestionReport";

describe("Question report", () => {
  let wrapper: ShallowWrapper;
  const question: Question = {
    questionTitle: "Question title",
    choices: [
      {
        text: "Choice 1",
        selections: 10,
      },
      {
        text: "Choice 2",
        selections: 30,
      },
    ],
  };

  beforeEach(() => {
    wrapper = shallow(<QuestionReport question={question}></QuestionReport>);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("renders question title", () => {
    const title = wrapper.find(".questionTitle");
    expect(title.text()).toBe(question.questionTitle);
  });

  it("renders choices with percentages", () => {
    const choices = wrapper.find(".choices").children();
    expect(choices.length).toBe(2);

    const choicesTexts = choices.map((choice) => choice.find("span"));
    expect(choicesTexts[0].text()).toBe(question.choices[0].text);
    expect(choicesTexts[1].text()).toBe(question.choices[1].text);

    const choicesPercentages = choices.map((choice) => choice.find("Progress"));
    expect(choicesPercentages[0].prop("percent")).toBe(25);
    expect(choicesPercentages[1].prop("percent")).toBe(75);
  });
});
