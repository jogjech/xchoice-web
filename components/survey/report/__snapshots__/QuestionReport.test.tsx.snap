// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question report renders correctly 1`] = `
<Fragment>
  <div>
    <p
      className="questionTitle"
    >
      Question title
    </p>
  </div>
  <div
    className="choices"
  >
    <div
      key="0"
    >
      <span>
        Choice 1
      </span>
      <Progress
        percent={25}
        showInfo={true}
        size="default"
        strokeLinecap="round"
        trailColor={null}
        type="line"
      />
    </div>
    <div
      key="1"
    >
      <span>
        Choice 2
      </span>
      <Progress
        percent={75}
        showInfo={true}
        size="default"
        strokeLinecap="round"
        trailColor={null}
        type="line"
      />
    </div>
  </div>
</Fragment>
`;
